File: gp/src/main/java/uk/healthtechwales/gppractice/service/ClamAVService.java
package uk.healthtechwales.gppractice.service;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.io.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.Base64;

@Service
public class ClamAVService {

    private static final Logger log = LoggerFactory.getLogger(ClamAVService.class);

    @Value("${clamav.host}")
    private String clamAvHost;

    @Value("${clamav.port}")
    private int clamAvPort;

    @Value("${clamav.timeout:5000}")
    private int timeout;

    /**
     * Scans a base64-encoded document for malware
     * @param base64Data the base64-encoded document
     * @return ScanResult containing scan status and details
     */
    public ScanResult scanDocument(String base64Data) {
        try {
            // Decode base64 to bytes
            byte[] documentBytes = Base64.getDecoder().decode(base64Data);
            return scanFile(documentBytes);
        } catch (Exception e) {
            log.error("Error decoding base64 or scanning document", e);
            return ScanResult.builder()
                    .status(ScanStatus.ERROR)
                    .virusName("Failed to scan document: " + e.getMessage())
                    .build();
        }
    }

    /**
     * Scan a file for viruses using ClamAV
     * @param fileData the file data to scan
     * @return ScanResult containing scan status and virus name if infected
     */
    public ScanResult scanFile(byte[] fileData) throws IOException {
        log.info("Connecting to ClamAV at {}:{}", clamAvHost, clamAvPort);
        
        try (Socket socket = new Socket(clamAvHost, clamAvPort)) {
            socket.setSoTimeout(timeout);
            
            try (OutputStream out = socket.getOutputStream();
                 InputStream in = socket.getInputStream()) {
                
                // Send INSTREAM command
                out.write("zINSTREAM\0".getBytes(StandardCharsets.UTF_8));
                out.flush();
                
                // Send file data in chunks
                int chunkSize = 2048;
                int offset = 0;
                
                while (offset < fileData.length) {
                    int length = Math.min(chunkSize, fileData.length - offset);
                    
                    // Send chunk size (4 bytes, network byte order)
                    byte[] sizeBytes = ByteBuffer.allocate(4).putInt(length).array();
                    out.write(sizeBytes);
                    
                    // Send chunk data
                    out.write(fileData, offset, length);
                    offset += length;
                }
                
                // Send zero-length chunk to indicate end of stream
                out.write(new byte[]{0, 0, 0, 0});
                out.flush();
                
                // Read response
                BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
                String response = reader.readLine();
                
                log.info("ClamAV response: {}", response);
                
                return parseResponse(response);
            }
        } catch (IOException e) {
            log.error("Error connecting to ClamAV", e);
            return ScanResult.builder()
                    .status(ScanStatus.ERROR)
                    .virusName("Failed to scan file with ClamAV: " + e.getMessage())
                    .build();
        }
    }

    /**
     * Parse ClamAV response
     * Expected responses:
     * - "stream: OK" - file is clean
     * - "stream: <virus name> FOUND" - virus detected
     */
    private ScanResult parseResponse(String response) {
        if (response == null || response.isEmpty()) {
            return ScanResult.builder()
                    .status(ScanStatus.ERROR)
                    .virusName("Unknown - empty response from ClamAV")
                    .build();
        }

        // Remove "stream: " prefix if present
        String result = response.replace("stream: ", "").trim();

        if (result.equals("OK")) {
            log.info("File is clean");
            return ScanResult.builder()
                    .status(ScanStatus.CLEAN)
                    .build();
        } else if (result.contains("FOUND")) {
            // Extract virus name
            String virusName = result.replace(" FOUND", "").trim();
            log.warn("Virus detected: {}", virusName);
            return ScanResult.builder()
                    .status(ScanStatus.INFECTED)
                    .virusName(virusName)
                    .build();
        } else {
            log.error("Unexpected ClamAV response: {}", response);
            return ScanResult.builder()
                    .status(ScanStatus.ERROR)
                    .virusName("Unexpected response: " + response)
                    .build();
        }
    }

    /**
     * Test connection to ClamAV by sending PING command
     */
    public boolean ping() {
        try (Socket socket = new Socket(clamAvHost, clamAvPort)) {
            socket.setSoTimeout(timeout);
            
            try (OutputStream out = socket.getOutputStream();
                 InputStream in = socket.getInputStream()) {
                
                out.write("zPING\0".getBytes(StandardCharsets.UTF_8));
                out.flush();
                
                BufferedReader reader = new BufferedReader(new InputStreamReader(in, StandardCharsets.UTF_8));
                String response = reader.readLine();
                
                log.debug("ClamAV PING response: {}", response);
                return "PONG".equals(response.replace("\0", "").trim());
            }
        } catch (IOException e) {
            log.error("Failed to ping ClamAV", e);
            return false;
        }
    }

    public enum ScanStatus {
        CLEAN,
        INFECTED,
        ERROR
    }

    public static class ScanResult {
        private ScanStatus status;
        private String virusName;

        public ScanResult() {}

        private ScanResult(Builder builder) {
            this.status = builder.status;
            this.virusName = builder.virusName;
        }

        public ScanStatus getStatus() {
            return status;
        }

        public void setStatus(ScanStatus status) {
            this.status = status;
        }

        public String getVirusName() {
            return virusName;
        }

        public void setVirusName(String virusName) {
            this.virusName = virusName;
        }

        public boolean isClean() {
            return status == ScanStatus.CLEAN;
        }

        public String getDetails() {
            if (status == ScanStatus.CLEAN) {
                return "stream: OK";
            } else if (status == ScanStatus.INFECTED) {
                return "stream: " + virusName + " FOUND";
            } else {
                return virusName != null ? virusName : "Unknown error";
            }
        }

        public static Builder builder() {
            return new Builder();
        }

        public static class Builder {
            private ScanStatus status;
            private String virusName;

            public Builder status(ScanStatus status) {
                this.status = status;
                return this;
            }

            public Builder virusName(String virusName) {
                this.virusName = virusName;
                return this;
            }

            public ScanResult build() {
                return new ScanResult(this);
            }
        }
    }
}

File: src/gp_practice_documents/services/clamav_service.py
import base64
import logging
import socket
import struct
from enum import Enum
from typing import Optional

from pydantic import BaseModel

from ..config import Settings

logger = logging.getLogger(__name__)


class ScanStatus(str, Enum):
    """ClamAV scan status."""
    CLEAN = "CLEAN"
    INFECTED = "INFECTED"
    ERROR = "ERROR"


class ScanResult(BaseModel):
    """Result of a ClamAV scan."""
    status: ScanStatus
    virus_name: Optional[str] = None

    def is_clean(self) -> bool:
        """Check if the scan result is clean."""
        return self.status == ScanStatus.CLEAN

    def get_details(self) -> str:
        """Get scan result details."""
        if self.status == ScanStatus.CLEAN:
            return "stream: OK"
        elif self.status == ScanStatus.INFECTED:
            return f"stream: {self.virus_name} FOUND"
        else:
            return self.virus_name or "Unknown error"


class ClamAVService:
    """Service for scanning documents with ClamAV."""

    def __init__(self, settings: Settings):
        self.host = settings.clamav_host
        self.port = settings.clamav_port
        self.timeout = settings.clamav_timeout / 1000.0  # Convert ms to seconds

    def scan_document(self, base64_data: str) -> ScanResult:
        """
        Scan a base64-encoded document for malware.

        Args:
            base64_data: Base64-encoded document data

        Returns:
            ScanResult containing scan status and details
        """
        try:
            # Decode base64 to bytes
            document_bytes = base64.b64decode(base64_data)
            return self.scan_file(document_bytes)
        except Exception as e:
            logger.error(f"Error decoding base64 or scanning document: {e}")
            return ScanResult(
                status=ScanStatus.ERROR,
                virus_name=f"Failed to scan document: {str(e)}"
            )

    def scan_file(self, file_data: bytes) -> ScanResult:
        """
        Scan a file for viruses using ClamAV.

        Args:
            file_data: File data to scan

        Returns:
            ScanResult containing scan status and virus name if infected
        """
        logger.info(f"Connecting to ClamAV at {self.host}:{self.port}")

        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(self.timeout)
                sock.connect((self.host, self.port))

                # Send INSTREAM command
                sock.sendall(b"zINSTREAM\0")

                # Send file data in chunks
                chunk_size = 2048
                offset = 0

                while offset < len(file_data):
                    length = min(chunk_size, len(file_data) - offset)

                    # Send chunk size (4 bytes, network byte order)
                    sock.sendall(struct.pack("!I", length))

                    # Send chunk data
                    sock.sendall(file_data[offset:offset + length])
                    offset += length

                # Send zero-length chunk to indicate end of stream
                sock.sendall(struct.pack("!I", 0))

                # Read response
                response = sock.recv(4096).decode("utf-8").strip()

                logger.info(f"ClamAV response: {response}")

                return self._parse_response(response)

        except Exception as e:
            logger.error(f"Error connecting to ClamAV: {e}")
            return ScanResult(
                status=ScanStatus.ERROR,
                virus_name=f"Failed to scan file with ClamAV: {str(e)}"
            )

    def _parse_response(self, response: str) -> ScanResult:
        """
        Parse ClamAV response.

        Expected responses:
        - "stream: OK" - file is clean
        - "stream: <virus name> FOUND" - virus detected
        """
        if not response:
            return ScanResult(
                status=ScanStatus.ERROR,
                virus_name="Unknown - empty response from ClamAV"
            )

        # Remove "stream: " prefix if present
        result = response.replace("stream: ", "").strip()

        if result == "OK":
            logger.info("File is clean")
            return ScanResult(status=ScanStatus.CLEAN)
        elif "FOUND" in result:
            # Extract virus name
            virus_name = result.replace(" FOUND", "").strip()
            logger.warning(f"Virus detected: {virus_name}")
            return ScanResult(
                status=ScanStatus.INFECTED,
                virus_name=virus_name
            )
        else:
            logger.error(f"Unexpected ClamAV response: {response}")
            return ScanResult(
                status=ScanStatus.ERROR,
                virus_name=f"Unexpected response: {response}"
            )

    def ping(self) -> bool:
        """Test connection to ClamAV by sending PING command."""
        try:
            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
                sock.settimeout(self.timeout)
                sock.connect((self.host, self.port))

                sock.sendall(b"zPING\0")

                response = sock.recv(4096).decode("utf-8").strip()

                logger.debug(f"ClamAV PING response: {response}")
                return response.replace("\0", "").strip() == "PONG"

        except Exception as e:
            logger.error(f"Failed to ping ClamAV: {e}")
            return False
